test/parallel/test-http2-ping.js

```diff
@@ -0,0 +1,72 @@
+'use strict';
+
+const common = require('../common');
+if (!common.hasCrypto)
+  common.skip('missing crypto');
+
+const async_hooks = require('async_hooks');
+const assert = require('assert');
+const http2 = require('http2');
+
+const pings = new Set();
+const events = [0, 0, 0, 0];
+
+const hook = async_hooks.createHook({
+  init(id, type, trigger, resource) {
+    if (type === 'HTTP2PING') {
+      pings.add(id);
+      events[0]++;
+    }
+  },
+  before(id) {
+    if (pings.has(id)) {
+      events[1]++;
+    }
+  },
+  after(id) {
+    if (pings.has(id)) {
+      events[2]++;
+    }
+  },
+  destroy(id) {
+    if (pings.has(id)) {
+      events[3]++;
+    }
+  }
+});
+hook.enable();
+
+process.on('exit', () => {
+  assert.deepStrictEqual(events, [3, 3, 3, 3]);
+});
+
+const server = http2.createServer();
+server.on('stream', common.mustCall((stream) => {
+  assert(stream.session.ping(common.mustCall((ack, duration) => {
+    assert(ack);
+    assert.strictEqual(typeof duration, 'number');
+    stream.end('ok');
+  })));
+  stream.respond();
+}));
+
+server.listen(0, common.mustCall(() => {
+  const client = http2.connect(`http://localhost:${server.address().port}`);
+  client.on('connect', common.mustCall(() => {
+    assert(client.ping(common.mustCall((ack, duration) => {
+      assert(ack);
+      assert.strictEqual(typeof duration, 'number');
+    })));
+    // Only one ping at a time
```

Comment:
Why only one ping at a time? I see no such limitation in the http2 spec.

Reply:
The protocol doesn't limit it, but PING is a potential DOS vector. The one at a time limit is intended to avoid the possibility of abuse.

Reply:
What risk does PING pose over sending any other outgoing REQUEST/SETTINGS/PRIORITY frame?

Multiple outstanding PINGs could occur as keepalive heartbeat messages under high RTT conditions.

Reply:
We also currently limit SETTINGS frame too, just not as strictly. The risk is in allowing node.js to be used inappropriately as an attack tool. It may be ok tho

Reply:
We could allow a configurable upper limit on the number of outstanding pings, allowing more than one to be sent at a time, and using a stack for the Http2Ping callbacks

Reply:
The FIFO approach should generally work, though, if I understand the spec correctly, there is a potential case where outstanding PING are not ACKed in the same order as the outgoing sequence. Methods for dealing with this would be limiting concurrency to 1, or matching the ACKed payload (assuming unique bytes).